src/main/kotlin/graphics/scenery/utils/TGAReader.kt

```diff
@@ -0,0 +1,630 @@
+                        decodeBuffer[decoded++] = elements[j]
+                    }
+                }
+            } else { // RAW
+                val count = (packet + 1) * elementCount
+                for(i in 0 until count) {
+                    decodeBuffer[decoded++] = buffer[offset++]
+                }
+            }
+        }
+        return decodeBuffer
+    }
+
+    @Throws(IOException::class)
+    private fun createPixelsFromColormap(
+        width: Int, height: Int, depth: Int,
+        bytes: ByteArray, offset: Int, palette: ByteArray, colormapOrigin: Int, descriptor: Int,
+        order: Order
+    ): IntArray {
+        var pixels: IntArray? = null
+        val rs = order.redShift
+        val gs = order.greenShift
+        val bs = order.blueShift
+        val `as` = order.alphaShift
+        when(depth) {
+            24 -> {
+                pixels = IntArray(width * height)
+                if(descriptor and RIGHT_ORIGIN != 0) {
+                    if(descriptor and UPPER_ORIGIN != 0) {
+                        // UpperRight
+                        var i = 0
+                        while(i < height) {
+                            var j = 0
+                            while(j < width) {
+                                val colormapIndex = bytes[offset + width * i + j].toInt() and 0xFF - colormapOrigin
+                                var color = -0x1
+                                if(colormapIndex >= 0) {
+                                    val index = 3 * colormapIndex + 18
+                                    val b = palette[index + 0].toInt() and 0xFF
+                                    val g = palette[index + 1].toInt() and 0xFF
+                                    val r = palette[index + 2].toInt() and 0xFF
+                                    val a = 0xFF
+                                    color = r shl rs or (g shl gs) or (b shl bs) or (a shl `as`)
+                                }
+                                pixels[width * i + (width - j - 1)] = color
+                                j++
+                            }
+                            i++
+                        }
+                    } else {
+                        // LowerRight
+                        var i = 0
+                        while(i < height) {
+                            var j = 0
+                            while(j < width) {
+                                val colormapIndex = bytes[offset + width * i + j].toInt() and 0xFF - colormapOrigin
+                                var color = -0x1
+                                if(colormapIndex >= 0) {
+                                    val index = 3 * colormapIndex + 18
+                                    val b = palette[index + 0].toInt() and 0xFF
+                                    val g = palette[index + 1].toInt() and 0xFF
+                                    val r = palette[index + 2].toInt() and 0xFF
+                                    val a = 0xFF
```

Comment:
There is a lot of repeated code here. You could factor out e.g. the BGR extraction to a helper method that returns a `Triple`.

Reply:
Won't do that, that code is just copied over from ClearGL and I don't care too much about it. We need it though, for the reason of being able to read TGA files.